File: user/src/lib/tls/ecdsa.zig
//! ECDSA verification over NIST P-256 and P-384 (FIPS 186-4 §6.4, verify
//! only). Freestanding, fixed capacity, no allocation.
//!
//! Field arithmetic is Montgomery multiplication modulo the curve prime
//! (`bigint.zig`); points use Jacobian coordinates for the scalar multiply and
//! affine for the result. Verification is on public data, so the ladder is
//! plain double-and-add — not constant-time — which is fine here and stated
//! rather than hidden.
//!
//! The curve constants are not typed by hand: they come from
//! `openssl ecparam -param_enc explicit` (see `ecdsa_vectors.zig`), and the
//! verifier is checked against `openssl dgst -sign` signatures.

const std = @import("std");
const bigint = @import("bigint.zig");
const crypto = @import("crypto");

const max_limbs = bigint.max_limbs;
const L = max_limbs;

const Fe = [L]u64;

fn zero() Fe {
    return [_]u64{0} ** L;
}

pub const Error = error{
    SignatureOutOfRange,
    NotOnCurve,
    PointAtInfinity,
    ValueTooLarge,
};

pub const Curve = struct {
    mont: bigint.Mont, // mod p
    n_mont: bigint.Mont, // mod n
    p: Fe, // plain p, for conditional subtraction
    n_plain: Fe, // plain n
    b: Fe, // b in Montgomery form
    gx: Fe, // in Montgomery form
    gy: Fe, // in Montgomery form
    nbytes: usize,

    pub fn init(p_bytes: []const u8, b_bytes: []const u8, n_bytes: []const u8, gx_bytes: []const u8, gy_bytes: []const u8) (bigint.Error || Error)!Curve {
        const mont = bigint.Mont.init(p_bytes) catch return Error.ValueTooLarge;
        const n_mont = bigint.Mont.init(n_bytes) catch return Error.ValueTooLarge;
        var self = Curve{
            .mont = mont,
            .n_mont = n_mont,
            .p = undefined,
            .n_plain = undefined,
            .b = undefined,
            .gx = undefined,
            .gy = undefined,
            .nbytes = n_bytes.len,
        };
        const pv = try bigint.Int.fromBytesBE(p_bytes);
        const nv = try bigint.Int.fromBytesBE(n_bytes);
        self.p = pv.limbs;
        self.n_plain = nv.limbs;
        self.b = self.toFe(b_bytes) catch unreachable;
        self.gx = self.toFe(gx_bytes) catch unreachable;
        self.gy = self.toFe(gy_bytes) catch unreachable;
        return self;
    }

    fn toFe(self: *const Curve, bytes: []const u8) (bigint.Error || Error)!Fe {
        const v = try bigint.Int.fromBytesBE(bytes);
        if (v.n > self.mont.n) return Error.ValueTooLarge;
        return self.mont.toMont(v.limbs[0..self.mont.n]);
    }

    // --- field arithmetic (Montgomery form) ---

    fn feAdd(self: *const Curve, a: Fe, b: Fe) Fe {
        // a, b < p so a+b < 2p. A carry into limb n means a+b >= 2^(64n), and in
        // that case subtracting p wraps back correctly (r + 2^(64n) - p == r - p
        // mod 2^(64n)), so both cases collapse to one conditional subtraction.
        var r = a;
        const carry = addInto(&r, &b, self.mont.n);
        if (carry != 0 or cmpSlice(r[0..self.mont.n], self.p[0..self.mont.n], self.mont.n) >= 0) {
            _ = subInto(&r, self.p[0..self.mont.n], self.mont.n);
        }
        return r;
    }

    fn feSub(self: *const Curve, a: Fe, b: Fe) Fe {
        var r = a;
        const borrow = subInto(&r, &b, self.mont.n);
        if (borrow != 0) _ = addInto(&r, self.p[0..self.mont.n], self.mont.n);
        return r;
    }

    fn feMul(self: *const Curve, a: Fe, b: Fe) Fe {
        return self.mont.mul(&a, &b);
    }

    fn feSquare(self: *const Curve, a: Fe) Fe {
        return self.mont.mul(&a, &a);
    }

    fn feIsZero(a: Fe, n: usize) bool {
        for (a[0..n]) |limb| if (limb != 0) return false;
        return true;
    }

    fn feCmp(a: Fe, b: Fe, n: usize) i8 {
        return cmpSlice(a[0..n], b[0..n], n);
    }

    fn feInv(self: *const Curve, a: Fe) Fe {
        // a^-1 mod p via Fermat. a is in Montgomery form; convert out, invert
        // with p-2, convert back.
        const plain = self.mont.fromMont(&a);
        var pi = bigint.Int{ .limbs = plain, .n = self.mont.n };
        pi.trim();
        var buf: [bigint.max_bytes]u8 = undefined;
        const bytes = pi.writeBytesBETrim(&buf);
        var e: [bigint.max_bytes]u8 = undefined;
        // exponent = p - 2
        var p_int = bigint.Int{ .limbs = self.p, .n = self.mont.n };
        p_int.trim();
        var exp = p_int.limbs;
        var borrow: u64 = 2;
        var i: usize = 0;
        while (borrow != 0 and i < self.mont.n) : (i += 1) {
            const s = @as(u128, exp[i]) -% @as(u128, borrow);
            exp[i] = @truncate(s);
            borrow = @truncate(s >> 127);
        }
        var eint = bigint.Int{ .limbs = exp, .n = self.mont.n };
        eint.trim();
        const elen = eint.writeBytesBETrim(&e);
        var inv: [max_limbs]u64 = undefined;
        self.mont.expMod(bytes, elen, &inv) catch unreachable;
        // expMod returns the plain inverse; re-enter the Montgomery domain.
        return self.mont.toMont(inv[0..self.mont.n]);
    }

    // --- points ---

    // Jacobian point: (X, Y, Z) plus an explicit infinity flag. Affine
    // results carry x/y with z left zero; only x/y/infinity are read there.
    const Point = struct { x: Fe = zero(), y: Fe = zero(), z: Fe = zero(), infinity: bool = true };

    fn isInfinity(p: Point) bool {
        return p.infinity;
    }

    /// Jacobian doubling (a = -3 optimized).
    fn jDouble(self: *const Curve, p: Point) Point {
        if (p.infinity) return p;
        if (feIsZero(p.y, self.mont.n)) return .{ .infinity = true };

        const a = feSquare(self, p.x);
        const b = feSquare(self, p.y);
        const c = feSquare(self, b);
        // D = 2*((X+B)^2 - A - C)
        var d = feSquare(self, feAdd(self, p.x, b));
        d = feSub(self, d, a);
        d = feSub(self, d, c);
        d = feAdd(self, d, d);
        // E = 3*A - 3*Z^4 (a = -3), with Z4 = Z^2 squared
        const z2 = feSquare(self, p.z);
        const z4 = feSquare(self, z2);
        var e = feAdd(self, a, a);
        e = feAdd(self, e, a); // 3A
        const three_z4 = feAdd(self, z4, feAdd(self, z4, z4));
        e = feSub(self, e, three_z4);
        const f = feSquare(self, e);
        // X3 = F - 2D
        const x3 = feSub(self, f, feAdd(self, d, d));
        // Y3 = E*(D - X3) - 8C
        var y3 = feSub(self, d, x3);
        y3 = feMul(self, e, y3);
        var c8 = feAdd(self, c, c); // 2C
        c8 = feAdd(self, c8, c8); // 4C
        c8 = feAdd(self, c8, c8); // 8C
        y3 = feSub(self, y3, c8);
        // Z3 = 2*Y*Z
        const z3 = feAdd(self, feMul(self, p.y, p.z), feMul(self, p.y, p.z));
        return .{ .x = x3, .y = y3, .z = z3, .infinity = false };
    }

    /// Mixed addition: Jacobian P + affine Q.
    fn jMixedAdd(self: *const Curve, p: Point, qx: Fe, qy: Fe) Point {
        if (p.infinity) return .{ .x = qx, .y = qy, .z = self.mont.one, .infinity = false };
        const z1z1 = feSquare(self, p.z);
        const u2x = feMul(self, qx, z1z1);
        const s2 = feMul(self, feMul(self, qy, p.z), z1z1);
        const h = feSub(self, u2x, p.x);
        const r = feSub(self, s2, p.y);
        if (feIsZero(h, self.mont.n)) {
            if (feIsZero(r, self.mont.n)) return self.jDouble(p);
            return .{ .infinity = true };
        }
        const hh = feSquare(self, h);
        const hhh = feMul(self, h, hh);
        const v = feMul(self, p.x, hh);
        var x3 = feSquare(self, r);
        x3 = feSub(self, x3, hhh);
        x3 = feSub(self, x3, feAdd(self, v, v));
        var y3 = feMul(self, r, feSub(self, v, x3));
        y3 = feSub(self, y3, feMul(self, p.y, hhh));
        const z3 = feMul(self, p.z, h);
        return .{ .x = x3, .y = y3, .z = z3, .infinity = false };
    }

    fn toAffine(self: *const Curve, p: Point) Error!Point {
        if (p.infinity) return error.PointAtInfinity;
        const zinv = self.feInv(p.z);
        const zinv2 = feSquare(self, zinv);
        const x = feMul(self, p.x, zinv2);
        const y = feMul(self, feMul(self, p.y, zinv2), zinv);
        return .{ .x = x, .y = y, .infinity = false };
    }

    /// Affine point addition (handles double and infinity).
    fn affAdd(self: *const Curve, p: Point, q: Point) Error!Point {
        if (p.infinity) return q;
        if (q.infinity) return p;
        if (feCmp(p.x, q.x, self.mont.n) == 0) {
            if (feCmp(p.y, q.y, self.mont.n) == 0) return self.affDouble(p);
            return error.PointAtInfinity;
        }
        const slope = feMul(self, feSub(self, q.y, p.y), self.feInv(feSub(self, q.x, p.x)));
        var x3 = feSquare(self, slope);
        x3 = feSub(self, x3, p.x);
        x3 = feSub(self, x3, q.x);
        var y3 = feMul(self, slope, feSub(self, p.x, x3));
        y3 = feSub(self, y3, p.y);
        return .{ .x = x3, .y = y3, .infinity = false };
    }

    fn affDouble(self: *const Curve, p: Point) Error!Point {
        if (p.infinity) return p;
        if (feIsZero(p.y, self.mont.n)) return error.PointAtInfinity;
        // a = -3: slope = 3(x^2 - 1) / 2y
        const x2 = feSquare(self, p.x);
        const one = self.mont.one;
        var num = feSub(self, x2, one);
        num = feAdd(self, feAdd(self, num, num), num); // 3*(x^2 - 1)
        const den = feAdd(self, p.y, p.y); // 2y
        const slope = feMul(self, num, self.feInv(den));
        var x3 = feSquare(self, slope);
        x3 = feSub(self, x3, feAdd(self, p.x, p.x));
        var y3 = feMul(self, slope, feSub(self, p.x, x3));
        y3 = feSub(self, y3, p.y);
        return .{ .x = x3, .y = y3, .infinity = false };
    }

    fn scalarMult(self: *const Curve, k_bytes: []const u8, px: Fe, py: Fe) (bigint.Error || Error)!Point {
        const k = try bigint.Int.fromBytesBE(k_bytes);
        var acc = Point{ .infinity = true };
        var li = k.n;
        var started = false;
        while (li > 0) {
            li -= 1;
            const limb = k.limbs[li];
            if (!started and limb == 0) continue;
            started = true;
            var bit: usize = 64;
            while (bit > 0) {
                bit -= 1;
                acc = self.jDouble(acc);
                if ((limb >> @intCast(bit)) & 1 != 0) acc = self.jMixedAdd(acc, px, py);
            }
        }
        return self.toAffine(acc);
    }

    fn onCurve(self: *const Curve, x: Fe, y: Fe) bool {
        const lhs = feSquare(self, y);
        const x2 = feSquare(self, x);
        const x3 = feMul(self, x2, x);
        // a = -3
        var rhs = feSub(self, x3, feAdd(self, feAdd(self, x, x), x));
        rhs = feAdd(self, rhs, self.b);
        return feCmp(lhs, rhs, self.mont.n) == 0;
    }

    pub fn verify(self: *const Curve, qx_bytes: []const u8, qy_bytes: []const u8, r_bytes: []const u8, s_bytes: []const u8, z_bytes: []const u8) bool {
        const r = bigint.Int.fromBytesBE(r_bytes) catch return false;
        const s = bigint.Int.fromBytesBE(s_bytes) catch return false;
        if (r.isZero() or s.isZero()) return false;
        const n_int = bigint.Int{ .limbs = self.n_plain, .n = self.n_mont.n };
        if (r.cmp(&n_int) >= 0 or s.cmp(&n_int) >= 0) return false;

        const qx = self.toFe(qx_bytes) catch return false;
        const qy = self.toFe(qy_bytes) catch return false;
        if (!self.onCurve(qx, qy)) return false;

        // w = s^-1 mod n (Fermat).
        var s_int = bigint.Int.fromBytesBE(s_bytes) catch return false;
        s_int.trim();
        var sbuf: [bigint.max_bytes]u8 = undefined;
        const s_trim = s_int.writeBytesBETrim(&sbuf);
        var w: [max_limbs]u64 = undefined;
        self.n_mont.invModPrime(s_trim, &w) catch return false;

        // u1 = z*w, u2 = r*w (mod n).
        const z = bigint.Int.fromBytesBE(z_bytes) catch return false;
        const r_int = bigint.Int.fromBytesBE(r_bytes) catch return false;
        const uu1 = self.nMul(z, w);
        const uu2 = self.nMul(r_int, w);

        var u1b: [bigint.max_bytes]u8 = undefined;
        var u2b: [bigint.max_bytes]u8 = undefined;
        var t1 = bigint.Int{ .limbs = uu1, .n = self.n_mont.n };
        var t2 = bigint.Int{ .limbs = uu2, .n = self.n_mont.n };
        t1.trim();
        t2.trim();
        const u1s = t1.writeBytesBETrim(&u1b);
        const u2s = t2.writeBytesBETrim(&u2b);

        const R1 = self.scalarMult(u1s, self.gx, self.gy) catch return false;
        const R2 = self.scalarMult(u2s, qx, qy) catch return false;
        const R = self.affAdd(R1, R2) catch return false;
        if (R.infinity) return false;

        // r == R.x mod n. R.x < p, and p ~ n, so one subtraction suffices.
        const xp = self.mont.fromMont(&R.x);
        var xr = bigint.Int{ .limbs = xp, .n = self.mont.n };
        xr.trim();
        while (xr.cmp(&n_int) >= 0) {
            _ = subInto(&xr.limbs, self.n_plain[0..self.n_mont.n], self.n_mont.n);
            xr.trim();
        }
        var xrb: [bigint.max_bytes]u8 = undefined;
        const xrs = xr.writeBytesBETrim(&xrb);
        var rbuf: [bigint.max_bytes]u8 = undefined;
        const r_trim = r.writeBytesBETrim(&rbuf);
        return std.mem.eql(u8, xrs, r_trim);
    }

    fn nMul(self: *const Curve, a: bigint.Int, b: [max_limbs]u64) [max_limbs]u64 {
        // (a * b) mod n, plain result. b is already reduced (< n).
        var bb = bigint.Int{ .limbs = b, .n = self.n_mont.n };
        bb.trim();
        const am = self.n_mont.toMont(a.limbs[0..self.n_mont.n]);
        const bm = self.n_mont.toMont(bb.limbs[0..self.n_mont.n]);
        const rm = self.n_mont.mul(&am, &bm);
        return self.n_mont.fromMont(&rm);
    }
};

fn addInto(a: *[L]u64, b: []const u64, n: usize) u64 {
    var carry: u64 = 0;
    for (0..n) |i| {
        const s = @as(u128, a[i]) + b[i] + carry;
        a[i] = @truncate(s);
        carry = @truncate(s >> 64);
    }
    return carry;
}

fn subInto(a: *[L]u64, b: []const u64, n: usize) u64 {
    var borrow: u64 = 0;
    for (0..n) |i| {
        const t = @as(u128, a[i]) -% @as(u128, b[i]) -% @as(u128, borrow);
        a[i] = @truncate(t);
        borrow = @truncate(t >> 127);
    }
    return borrow;
}

fn cmpSlice(a: []const u64, b: []const u64, n: usize) i8 {
    var i = n;
    while (i > 0) {
        i -= 1;
        if (a[i] != b[i]) return if (a[i] > b[i]) 1 else -1;
    }
    return 0;
}

const vectors = @import("ecdsa_vectors.zig");

fn decode(out: []u8, hexstr: []const u8) usize {
    _ = std.fmt.hexToBytes(out[0 .. hexstr.len / 2], hexstr) catch |e| {
        std.debug.print("decode failed: hexlen={d} outlen={d} err={any}\n", .{ hexstr.len, hexstr.len / 2, e });
        unreachable;
    };
    return hexstr.len / 2;
}

test "ecdsa: verifies every OpenSSL signature and rejects mutations" {
    var p: [128]u8 = undefined;
    var b: [128]u8 = undefined;
    var nn: [128]u8 = undefined;
    var gx: [128]u8 = undefined;
    var gy: [128]u8 = undefined;
    var qx: [128]u8 = undefined;
    var qy: [128]u8 = undefined;
    var r: [128]u8 = undefined;
    var s: [128]u8 = undefined;
    var z: [128]u8 = undefined;

    for (vectors.sigs) |v| {
        const params: vectors.Curves = if (std.mem.eql(u8, v.curve, "prime256v1")) vectors.p256 else vectors.p384;
        const plen = decode(&p, params.p);
        const blen = decode(&b, params.b);
        const nlen = decode(&nn, params.n);
        const glen = decode(&gx, params.gx);
        const gylen = decode(&gy, params.gy);
        var curve = Curve.init(p[0..plen], b[0..blen], nn[0..nlen], gx[0..glen], gy[0..gylen]) catch unreachable;

        const qxl = decode(&qx, v.pubx);
        const qyl = decode(&qy, v.puby);
        const rl = decode(&r, v.r);
        const sl = decode(&s, v.s);

        // z = the message hash, truncated to the curve order length.
        const hlen: usize = if (std.mem.eql(u8, v.hash, "sha256")) 32 else 48;
        if (std.mem.eql(u8, v.hash, "sha256")) {
            crypto.sha256.sha256(z[0..32], v.msg);
        } else {
            crypto.sha384.sha384(z[0..48], v.msg);
        }
        const zlen = if (hlen > nlen) nlen else hlen;

        if (!curve.verify(qx[0..qxl], qy[0..qyl], r[0..rl], s[0..sl], z[0..zlen])) {
            std.debug.print("signature rejected: {s}\n", .{v.label});
            return error.SignatureRejected;
        }

        // Mutated signature bit must fail.
        var bad_s = s;
        bad_s[sl - 1] ^= 0x01;
        if (curve.verify(qx[0..qxl], qy[0..qyl], r[0..rl], bad_s[0..sl], z[0..zlen])) {
            return error.MutatedSignatureAccepted;
        }
        // Mutated message must fail.
        var m2: [128]u8 = undefined;
        @memcpy(m2[0..v.msg.len], v.msg);
        m2[0] ^= 0x01;
        var z2: [128]u8 = undefined;
        if (std.mem.eql(u8, v.hash, "sha256")) {
            crypto.sha256.sha256(z2[0..32], m2[0..v.msg.len]);
        } else {
            crypto.sha384.sha384(z2[0..48], m2[0..v.msg.len]);
        }
        if (curve.verify(qx[0..qxl], qy[0..qyl], r[0..rl], s[0..sl], z2[0..zlen])) {
            return error.MutatedMessageAccepted;
        }
    }
}

test "ecdsa: r=0, s=0, r>=n and an off-curve point are refused" {
    const params = vectors.p256;
    var p: [128]u8 = undefined;
    var b: [128]u8 = undefined;
    var nn: [128]u8 = undefined;
    var gx: [128]u8 = undefined;
    var gy: [128]u8 = undefined;
    const plen = decode(&p, params.p);
    const blen = decode(&b, params.b);
    const nlen = decode(&nn, params.n);
    const glen = decode(&gx, params.gx);
    const gylen = decode(&gy, params.gy);
    var curve = Curve.init(p[0..plen], b[0..blen], nn[0..nlen], gx[0..glen], gy[0..gylen]) catch unreachable;

    const v = vectors.sigs[0];
    var qx: [128]u8 = undefined;
    var qy: [128]u8 = undefined;
    var r: [128]u8 = undefined;
    var s: [128]u8 = undefined;
    var z: [128]u8 = undefined;
    const qxl = decode(&qx, v.pubx);
    const qyl = decode(&qy, v.puby);
    const rl = decode(&r, v.r);
    const sl = decode(&s, v.s);
    crypto.sha256.sha256(z[0..32], v.msg);

    const z0 = [_]u8{0x00};
    try std.testing.expect(!curve.verify(qx[0..qxl], qy[0..qyl], &z0, s[0..sl], z[0..32]));
    try std.testing.expect(!curve.verify(qx[0..qxl], qy[0..qyl], r[0..rl], &z0, z[0..32]));

    // r = n is out of range.
    try std.testing.expect(!curve.verify(qx[0..qxl], qy[0..qyl], nn[0..nlen], s[0..sl], z[0..32]));

    // An off-curve public key (qy + 1) is refused.
    var qy_bad = qy;
    qy_bad[qyl - 1] ^= 0x01;
    try std.testing.expect(!curve.verify(qx[0..qxl], qy_bad[0..qyl], r[0..rl], s[0..sl], z[0..32]));
}

File: user/go/tls/ecdsa.go
// ECDSA verification (FIPS 186-4 §4.1.4) over P-256 and P-384, on the
// Montgomery core from mont.go. Curve constants come from the generated
// vectors (curveP256params / curveP384params — OpenSSL `ecparam -param_enc
// explicit -text` origin), never typed by hand. Signature pins are the
// OpenSSL-generated vectors, which include mutated cases that must fail.
//
// All curve arithmetic runs on public data, so per ADR 0029's declared limits
// this is not constant-time.

package tls

// ecCurve is a Weierstrass curve with a = -3 (P-256 and P-384 both).
// Montgomery-form discipline: every field value stored is v·R mod p; montMul,
// addMod and subMod all preserve that form.
type ecCurve struct {
	f   *montCtx // field arithmetic mod p
	o   *montCtx // scalar arithmetic mod n
	bm  []uint64 // curve b, Montgomery form
	gx  []uint64 // base point, Montgomery affine
	gy  []uint64
	c3  []uint64 // Montgomery(3), for the a=-3 doubling formulas
	c4  []uint64 // Montgomery(4)
	c8  []uint64 // Montgomery(8)
	pBE []byte   // p in big-endian bytes (the affine coordinate width)
}

func newECCurve(params curveParams) (*ecCurve, error) {
	pBytes := hexDecode(params.p)
	// The generated constants carry openssl's leading sign byte; the curve
	// coordinate width is the minimal big-endian encoding.
	z := 0
	for z < len(pBytes)-1 && pBytes[z] == 0 {
		z++
	}
	pBytes = pBytes[z:]
	f, err := newMont(pBytes)
	if err != nil {
		return nil, err
	}
	o, err := newMont(hexDecode(params.n))
	if err != nil {
		return nil, err
	}
	return &ecCurve{
		f:   f,
		o:   o,
		bm:  f.toMont(f.fromBE(hexDecode(params.b))),
		gx:  f.toMont(f.fromBE(hexDecode(params.gx))),
		gy:  f.toMont(f.fromBE(hexDecode(params.gy))),
		c3:  f.toMont(montOne(f, 3)),
		c4:  f.toMont(montOne(f, 4)),
		c8:  f.toMont(montOne(f, 8)),
		pBE: pBytes,
	}, nil
}

func montOne(f *montCtx, v uint64) []uint64 {
	one := make([]uint64, f.nlimbs)
	one[0] = v
	return one
}

func hexDecode(s string) []byte {
	out := make([]byte, len(s)/2)
	for i := 0; i < len(out); i++ {
		out[i] = hexVal(s[2*i])<<4 | hexVal(s[2*i+1])
	}
	return out
}

func hexVal(c byte) byte {
	switch {
	case c >= '0' && c <= '9':
		return c - '0'
	case c >= 'a' && c <= 'f':
		return c - 'a' + 10
	case c >= 'A' && c <= 'F':
		return c - 'A' + 10
	}
	return 0
}

// ecJacobian is a point in Jacobian coordinates over the Montgomery field;
// an all-zero z is the point at infinity.
type ecJacobian struct {
	x, y, z []uint64
}

func (c *ecCurve) clone(p *ecJacobian) *ecJacobian {
	return &ecJacobian{
		x: append([]uint64{}, p.x...),
		y: append([]uint64{}, p.y...),
		z: append([]uint64{}, p.z...),
	}
}

func (c *ecCurve) infinity() *ecJacobian {
	return &ecJacobian{
		x: make([]uint64, c.f.nlimbs),
		y: make([]uint64, c.f.nlimbs),
		z: make([]uint64, c.f.nlimbs),
	}
}

// ecDouble implements the a = -3 Jacobian doubling formulas.
func (c *ecCurve) ecDouble(p *ecJacobian) *ecJacobian {
	if isZeroLimb(p.z) {
		return c.infinity()
	}
	f := c.f
	delta := f.montMul(p.z, p.z)
	gamma := f.montMul(p.y, p.y)
	beta := f.montMul(p.x, gamma)
	alpha := f.montMul(f.montMul(f.subMod(p.x, delta), f.addMod(p.x, delta)), c.c3)
	a2 := f.montMul(alpha, alpha)
	eightBeta := f.montMul(beta, c.c8)
	x3 := f.subMod(a2, eightBeta)
	yz := f.addMod(p.y, p.z)
	z3 := f.montMul(yz, yz)
	z3 = f.subMod(z3, gamma)
	z3 = f.subMod(z3, delta)
	fourBeta := f.montMul(beta, c.c4)
	y3 := f.subMod(fourBeta, x3)
	y3 = f.montMul(alpha, y3)
	gamma2 := f.montMul(gamma, gamma)
	y3 = f.subMod(y3, f.montMul(gamma2, c.c8))
	return &ecJacobian{x: x3, y: y3, z: z3}
}

// ecAdd is general Jacobian addition.
func (c *ecCurve) ecAdd(p, q *ecJacobian) *ecJacobian {
	if isZeroLimb(p.z) {
		return c.clone(q)
	}
	if isZeroLimb(q.z) {
		return c.clone(p)
	}
	f := c.f
	z1z1 := f.montMul(p.z, p.z)
	z2z2 := f.montMul(q.z, q.z)
	u1 := f.montMul(p.x, z2z2)
	u2 := f.montMul(q.x, z1z1)
	s1 := f.montMul(f.montMul(p.y, q.z), z2z2)
	s2 := f.montMul(f.montMul(q.y, p.z), z1z1)
	h := f.subMod(u2, u1)
	r := f.subMod(s2, s1)
	if isZeroLimb(h) {
		if isZeroLimb(r) {
			return c.ecDouble(p)
		}
		return c.infinity()
	}
	hh := f.montMul(h, h)
	hhh := f.montMul(h, hh)
	v := f.montMul(u1, hh)
	x3 := f.subMod(f.montMul(r, r), hhh)
	x3 = f.subMod(x3, v)
	x3 = f.subMod(x3, v)
	y3 := f.subMod(v, x3)
	y3 = f.montMul(r, y3)
	y3 = f.subMod(y3, f.montMul(s1, hhh))
	z3 := f.montMul(f.montMul(p.z, q.z), h)
	return &ecJacobian{x: x3, y: y3, z: z3}
}

// ecScalarMul computes k*P, MSB-first double-and-add (public data).
func (c *ecCurve) ecScalarMul(k []uint64, p *ecJacobian) *ecJacobian {
	hb := -1
	for i := len(k) - 1; i >= 0 && hb < 0; i-- {
		for bit := 63; bit >= 0; bit-- {
			if (k[i]>>uint(bit))&1 != 0 {
				hb = i*64 + bit
				break
			}
		}
	}
	res := c.infinity()
	if hb < 0 {
		return res
	}
	addend := c.clone(p)
	for i := hb; i >= 0; i-- {
		res = c.ecDouble(res)
		if (k[i/64]>>uint(i%64))&1 != 0 {
			res = c.ecAdd(res, addend)
		}
	}
	return res
}

// ecAffine converts a Jacobian point to affine Montgomery (x, y).
func (c *ecCurve) ecAffine(p *ecJacobian) ([]uint64, []uint64, bool) {
	if isZeroLimb(p.z) {
		return nil, nil, false
	}
	f := c.f
	ziM := f.toMont(f.invMod(f.fromMont(p.z)))
	zi2M := f.montMul(ziM, ziM)
	zi3M := f.montMul(zi2M, ziM)
	return f.montMul(p.x, zi2M), f.montMul(p.y, zi3M), true
}

// verifyECDSA checks an (r, s) signature over digest z against the affine
// public point (qx, qy). The digest length must match the curve coordinate
// width (SHA-256 for P-256, SHA-384 for P-384), so the "leftmost order bits"
// truncation of FIPS 186-4 4.1.4 is the straight decode.
func (c *ecCurve) verifyECDSA(qx, qy, rBig, sBig, z []byte) bool {
	f, o := c.f, c.o
	r := o.fromBE(rBig)
	s := o.fromBE(sBig)
	if o.cmp(r, o.n) >= 0 || o.cmp(s, o.n) >= 0 {
		return false
	}
	if isZeroLimb(r) || isZeroLimb(s) {
		return false
	}
	if len(z) != len(c.pBE) || len(qx) != len(c.pBE) || len(qy) != len(c.pBE) {
		return false
	}
	zz := o.fromBE(z)
	w := o.invMod(s)
	u1 := o.mulMod(zz, w)
	u2 := o.mulMod(r, w)
	p1 := c.ecScalarMul(u1, &ecJacobian{x: c.gx, y: c.gy, z: f.toMont(montOne(f, 1))})
	p2 := c.ecScalarMul(u2, &ecJacobian{
		x: f.toMont(f.fromBE(qx)),
		y: f.toMont(f.fromBE(qy)),
		z: f.toMont(montOne(f, 1)),
	})
	sum := c.ecAdd(p1, p2)
	xx, _, ok := c.ecAffine(sum)
	if !ok {
		return false
	}
	// v = x_aff mod n. p < 2n for both curves, so one conditional subtract
	// of n is an exact reduction (subMod re-adds n on borrow, which is the
	// identity when v < n already).
	plain := f.fromMont(xx)
	v := o.subMod(plain, o.n)
	return o.cmp(v, r) == 0
}
